### PR TITLE
[Refactor] Refactor FCAF3D

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        flip_ratio_bev_horizontal=0.5,
+        flip_ratio_bev_vertical=0.5),
+    dict(
+        type='GlobalRotScaleTrans',
+        rot_range=[-0.087266, 0.087266],
+        scale_ratio_range=[.9, 1.1],
+        translation_std=[.1, .1, .1],
+        shift_height=False),
+    dict(type='NormalizePointsColor', color_mean=None),
+    # dict(type='DefaultFormatBundle3D', class_names=class_names),
```
Comment (ZCMax) [inline configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py:30]: remove the comments

File: configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    # dict(type='Collect3D', keys=['points', 'gt_bboxes_3d', 'gt_labels_3d'])
```
Comment (ZCMax) [inline configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py:31]: same here

File: configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+]
+train_dataloader = dict(
+    dataset=dict(
+        type='RepeatDataset',
+        times=10,
+        dataset=dict(pipeline=train_pipeline, filter_empty_gt=True)))
+val_dataloader = dict(dataset=dict(pipeline=test_pipeline))
+test_dataloader = val_dataloader
+
+# optimizer = dict(type='AdamW', lr=0.001, weight_decay=0.0001)
```
Comment (ZCMax) [inline configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py:74]: remove these old config

File: configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    type='MultiStepLR',
+    begin=0,
+    end=12,
+    by_epoch=True,
+    milestones=[8, 11],
+    gamma=0.1)
+
+custom_hooks = [dict(type='EmptyCacheHook', after_iter=True)]
+
+# training schedule for 2x
```
Comment (ZCMax) [inline configs/fcaf3d/fcaf3d_8xb2_scannet-3d-18class.py:95]: 1x ?

File: configs/fcaf3d/fcaf3d_8xb2_sunrgbd-3d-10class.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    dict(type='LoadAnnotations3D'),
+    dict(type='PointSample', num_points=n_points),
+    dict(type='RandomFlip3D', sync_2d=False, flip_ratio_bev_horizontal=0.5),
+    dict(
+        type='GlobalRotScaleTrans',
+        rot_range=[-0.523599, 0.523599],
+        scale_ratio_range=[0.85, 1.15],
+        translation_std=[.1, .1, .1],
+        shift_height=False),
+    dict(type='DefaultFormatBundle3D', class_names=class_names),
```
Comment (ZCMax) [inline configs/fcaf3d/fcaf3d_8xb2_sunrgbd-3d-10class.py:28]: delete this

File: configs/fcaf3d/fcaf3d_8xb2_sunrgbd-3d-10class.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                type='RandomFlip3D',
+                sync_2d=False,
+                flip_ratio_bev_horizontal=0.5,
+                flip_ratio_bev_vertical=0.5),
+            dict(type='PointSample', num_points=n_points),
+            dict(
+                type='DefaultFormatBundle3D',
+                class_names=class_names,
+                with_label=False),
+            dict(type='Collect3D', keys=['points'])
```
Comment (ZCMax) [inline configs/fcaf3d/fcaf3d_8xb2_sunrgbd-3d-10class.py:59]: seems this config is not refactored

File: mmdet3d/models/backbones/mink_resnet.py [hunk 7]
```diff
@@ -58,7 +58,7 @@ def __init__(self, depth, in_channels, num_stages=4, pool=True):
 
         for i, num_blocks in enumerate(stage_blocks):
             setattr(
-                self, f'layer{i}',
+                self, f'layer{i + 1}',
```
Comment (ZCMax) [inline mmdet3d/models/backbones/mink_resnet.py:61]: why add 1 for layer?

File: mmdet3d/models/detectors/mink_single_stage.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        # self.head = self.bbox_head
+
+    def extract_feat(
+        self, batch_inputs_dict: Dict[str, Tensor]
+    ) -> Union[Tuple[torch.Tensor], Dict[str, Tensor]]:
+        """Directly extract features from the backbone+neck.
+
+        Args:
+            batch_inputs_dict (dict): The model input dict which include
+                'points', 'img' keys.
```
Comment (ZCMax) [inline mmdet3d/models/detectors/mink_single_stage.py:69]: remove `img` in docstring

File: mmdet3d/models/backbones/mink_resnet.py (hunk 7 above)
Comment (JingweiZhang12) [reply mmdet3d/models/backbones/mink_resnet.py:61]: The original implement in the master branch is `layer{i+1}`, which starts with 1 and we changed it in the refactoring process.  If we don't do this, the checkpoint in the README will not be loaded correctly.

File: mmdet3d/models/dense_heads/fcaf3d_head.py [hunk 9]
```diff
@@ -0,0 +1,667 @@
...
+        bbox_loss (dict, optional): Config of bbox loss.
+        cls_loss (dict, optional): Config of classification loss.
+        train_cfg (dict, optional): Config for train stage. Defaults to None.
+        test_cfg (dict, optional): Config for test stage. Defaults to None.
+        init_cfg (dict, optional): Config for weight initialization.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 n_classes,
```
Comment (ZCMax) [inline mmdet3d/models/dense_heads/fcaf3d_head.py:51]: num_classes

File: mmdet3d/models/dense_heads/fcaf3d_head.py [hunk 10]
```diff
@@ -0,0 +1,667 @@
...
+                 in_channels,
+                 out_channels,
+                 n_reg_outs,
```
Comment (ZCMax) [inline mmdet3d/models/dense_heads/fcaf3d_head.py:54]: num_reg_outs

File: mmdet3d/models/dense_heads/fcaf3d_head.py [hunk 11]
```diff
@@ -0,0 +1,667 @@
...
+                 voxel_size,
+                 pts_prune_threshold,
+                 pts_assign_threshold,
+                 pts_center_threshold,
+                 center_loss=dict(type='CrossEntropyLoss', use_sigmoid=True),
```
Comment (ZCMax) [inline mmdet3d/models/dense_heads/fcaf3d_head.py:59]: mmdet.CrossEntropyLoss

File: mmdet3d/models/dense_heads/fcaf3d_head.py [hunk 12]
```diff
@@ -0,0 +1,667 @@
...
+                 bbox_loss=dict(type='AxisAlignedIoULoss'),
+                 cls_loss=dict(type='FocalLoss'),
```
Comment (ZCMax) [inline mmdet3d/models/dense_heads/fcaf3d_head.py:61]: mmdet.FocalLoss

File: mmdet3d/models/dense_heads/fcaf3d_head.py [hunk 13]
```diff
@@ -0,0 +1,667 @@
...
+        self.pts_center_threshold = pts_center_threshold
+        self.center_loss = build_loss(center_loss)
+        self.bbox_loss = build_loss(bbox_loss)
+        self.cls_loss = build_loss(cls_loss)
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+        self._init_layers(in_channels, out_channels, n_reg_outs, n_classes)
+
+    @staticmethod
+    def _make_block(in_channels, out_channels):
```
Comment (ZCMax) [inline mmdet3d/models/dense_heads/fcaf3d_head.py:78]: add type-hint

File: mmdet3d/models/dense_heads/fcaf3d_head.py [hunk 14]
```diff
@@ -0,0 +1,667 @@
...
+        Returns:
+            torch.nn.Module: With corresponding layers.
+        """
+        return nn.Sequential(
+            ME.MinkowskiConvolution(
+                in_channels, out_channels, kernel_size=3, dimension=3),
+            ME.MinkowskiBatchNorm(out_channels), ME.MinkowskiELU())
+
+    @staticmethod
+    def _make_up_block(in_channels, out_channels):
```
Comment (ZCMax) [inline mmdet3d/models/dense_heads/fcaf3d_head.py:93]: add type-hint

Comment (ZCMax) [reply mmdet3d/models/dense_heads/fcaf3d_head.py:93]: please check other places

File: mmdet3d/models/dense_heads/fcaf3d_head.py [hunk 15]
```diff
@@ -0,0 +1,667 @@
...
+            center_pred, bbox_pred, cls_pred, point, prune_score = \
+                self._forward_single(out, self.scales[i])
+            center_preds.append(center_pred)
+            bbox_preds.append(bbox_pred)
+            cls_preds.append(cls_pred)
+            points.append(point)
+        return center_preds[::-1], bbox_preds[::-1], cls_preds[::-1], \
+            points[::-1]
+
+    # def forward_train(self, x, gt_bboxes, gt_labels, input_metas):
```
Comment (ZCMax) [inline mmdet3d/models/dense_heads/fcaf3d_head.py:178]: remove these commented codes

File: mmdet3d/models/detectors/mink_single_stage.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    pass
+
+from mmdet3d.registry import MODELS
+from mmdet3d.utils import ConfigType, OptConfigType, OptMultiConfig
+from .single_stage import SingleStage3DDetector
+
+
+@MODELS.register_module()
+class MinkSingleStage3DDetector(SingleStage3DDetector):
+    """MinkSingleStage3DDetector.
```
Comment (ZCMax) [inline mmdet3d/models/detectors/mink_single_stage.py:21]: r"""

File: mmdet3d/models/losses/rotated_iou_loss.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+from mmcv.ops import diff_iou_rotated_3d
+from torch import Tensor
+from torch import nn as nn
+
+from mmdet3d.registry import MODELS
+from mmdet.models.losses.utils import weighted_loss
+
+
+@weighted_loss
+def rotated_iou_3d_loss(pred, target):
```
Comment (ZCMax) [inline mmdet3d/models/losses/rotated_iou_loss.py:14]: add type-hint

**Review by ZCMax — APPROVED** [review]

**Comment by ZCMax** [comment]
LGTM, just need to fix the Circle CI

**Review by VVsssssk — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**